File: bindings/node/dist-tools/samples/registry_query_sample.js
// SPDX-License-Identifier: MPL-2.0
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const assert = require('node:assert/strict');
const { once } = require('node:events');
const net = require('node:net');
const zlink = require('../dist');
async function reservePort() {
    const server = net.createServer();
    server.listen(0, '127.0.0.1');
    await once(server, 'listening');
    const { port } = server.address();
    await new Promise((resolve, reject) => server.close((error) => error ? reject(error) : resolve()));
    return port;
}
async function waitForTopologyEntry(query, serviceName, endpoint) {
    const deadline = Date.now() + 5000;
    while (Date.now() < deadline) {
        const entry = query.snapshot({ serviceName }).find((item) => item.endpoint === endpoint);
        if (entry) {
            return entry;
        }
        await new Promise((resolve) => setImmediate(resolve));
    }
    return null;
}
async function main() {
    const ctx = new zlink.Context();
    const registry = new zlink.Registry(ctx);
    const query = new zlink.RegistryQueryClient(ctx);
    const discovery = new zlink.Discovery(ctx, zlink.ServiceType.SPOT, 'topology-entry-found');
    const node = new zlink.SpotNode(ctx);
    const pubEndpoint = `tcp://127.0.0.1:${await reservePort()}`;
    const routerEndpoint = `tcp://127.0.0.1:${await reservePort()}`;
    const serviceEndpoint = `tcp://127.0.0.1:${await reservePort()}`;
    try {
        registry.bind(pubEndpoint, routerEndpoint);
        query.connect(routerEndpoint);
        discovery.connectRegistry(routerEndpoint);
        node.attachDiscovery(discovery);
        node.bind(serviceEndpoint);
        const entry = await waitForTopologyEntry(query, 'topology-entry-found', serviceEndpoint);
        assert.ok(entry);
        assert.equal(entry.endpoint, serviceEndpoint);
        console.log('[registry-query] connect → snapshot: "topology-entry-found"');
    }
    finally {
        discovery.close();
        query.close();
        registry.close();
        ctx.close();
    }
}
main().catch((error) => {
    console.error(error);
    process.exitCode = 1;
});

File: bindings/java/samples/Zlink.Samples/src/main/java/dev/kairoscode/zlink/samples/RegistryQuerySample.java
package dev.kairoscode.zlink.samples;

import dev.kairoscode.zlink.Context;
import dev.kairoscode.zlink.PubSocket;
import dev.kairoscode.zlink.service.discovery.Discovery;
import dev.kairoscode.zlink.service.registry.Registry;
import dev.kairoscode.zlink.service.registry.RegistryQueryClient;
import dev.kairoscode.zlink.service.registry.AutoConnectType;

public final class RegistryQuerySample {
    public static void main(String[] args) {
        SampleSupport.ensureNative();
        String serviceName = "sample";
        String registryPub = SampleSupport.tcpEndpoint();
        String registryRouter = SampleSupport.tcpEndpoint();
        String serviceEndpoint = SampleSupport.tcpEndpoint();
        Context ctx = new Context();
        Registry registry = null;
        Discovery discovery = null;
        RegistryQueryClient query = null;
        PubSocket provider = null;
        try {
            registry = new Registry(ctx);
            discovery = new Discovery(ctx, AutoConnectType.FANOUT, serviceName);
            query = new RegistryQueryClient(ctx);
            provider = new PubSocket(ctx);
            registry.bind(registryPub, registryRouter);
            discovery.connectRegistry(registryRouter);
            provider.attachDiscovery(discovery);
            provider.bind(serviceEndpoint);
            query.connect(registryRouter);
            final RegistryQueryClient finalQuery = query;

            SampleSupport.waitUntil("registry query sample",
                () -> {
                    try {
                        return finalQuery.snapshot().stream()
                            .anyMatch(entry -> serviceName.equals(entry.channelName()));
                    } catch (RuntimeException ex) {
                        return false;
                    }
                });

            System.out.println("[registry-query] service: \"sample\" -> snapshot: found");
        } finally {
            SampleSupport.closeQuietly(provider);
            SampleSupport.closeQuietly(query);
            SampleSupport.closeQuietly(discovery);
            SampleSupport.closeQuietly(registry);
            SampleSupport.closeQuietly(ctx);
        }
    }
}
